### ISSUE TITLE
Rule request: [no-async-without-await] Functions marked async must contain an await or return statement

### ISSUE THREAD

**Issue by pm-dev** [realm/SwiftLint#5082]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request
Disallow `async` functions which have no `await` expression.

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.
`async` functions that don't suspend contradict the [Swift Concurrency documentation](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/#Defining-and-Calling-Asynchronous-Functions) which specifies they are able to suspend partway through execution. 
Asynchronous functions that don’t use await can cause confusion in that a programmer awaiting such a function may assume it's guaranteed to execute its work in the background. Although such an assumption is a fundamental misunderstanding of Swift Concurrency, there are plenty of [examples](https://forums.swift.org/t/when-should-you-mark-a-function-async/64898/2).

Async functions without await are a common and unintentional result of refactoring.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

Pass:
```swift
func f() async {
    await fetch();
}
```

Fail:
```swift
func add(a: Int, b: Int) async {
    return a + b
}
```

4. Should the rule be configurable, if so what parameters should be configurable?
There may be edge cases where async without await is appropriate that I'm not thinking of.

6. Should the rule be opt-in or enabled by default? Why?
Enabled by default
I believe this rule would be fast, is not prone to false positives and is general consensus.

Linters for other languages have this rule:
[TS Lint](https://palantir.github.io/tslint/rules/no-async-without-await/)
[ES Lint](https://eslint.org/docs/latest/rules/require-await)

**Comment by pm-dev** [comment]
@jshier you're suggesting `async` can be added to run the function in the background, yet admit asynchronous functions aren't guaranteed to run in the background. This is why I believe this rule is important, to avoid this confusion.

Until Swift 5.9 and [custom executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md), it's not possible to use Swift Concurrency to guarantee code is run in the background.

**Comment by jshier** [comment]
If defined outside an actor context, `async` does guarantee background execution (execution on the default executor, i.e. the concurrency thread pool). If defined in an actor context, `async` still forces concurrent execution back onto the same context. Both of these are useful behaviors. Even in an actor context where the method will execute on the actor, this warning doesn't seem useful, as that fact should be self evident if you're familiar with Swift concurrency (e.g. unlabeled async functions will execute on their inherited actor context, if any, or the default executor), so I'm not sure what it would be protecting against.

**Comment by pm-dev** [comment]
Execution on the default executor != execution in the background. Today that may be the case, but the way the cooperative thread pool works is an implementation detail that is not exposed and could change in the future/on more constrained systems.  Swift Concurrency provides an abstraction that doesn't have a concept of "background". Currently (Until swift 5.9) if you want to guarantee code is executed off the main thread (i.e. the background) you need to drop down to GCD or Threads.

If this rule is not useful, then why shouldn't every function be marked as `async`?

**Comment by jshier** [comment]
Yeah, that's not a distinction Swift concurrency makes, so I think you're pretty off base on user expectations here. Why would anyone expect that when, given your definition, it's not even possible right now?

In any case, I've laid out my reasoning, it's up to the maintainers now.

**Comment by pm-dev** [comment]
> that's not a distinction Swift concurrency makes
I'd argue it does make that distinction by intentionally not documenting whether code is run in the background:
```
Swift's concurrency design is intentionally vague about the details of how code is actually run.
```
[Source](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md#motivation)

My response to your original comment only intended to explain why the one example you provided against this rule is invalid. Adding `async` to a function without awaits is not a way to run synchronous code in the background.

**Comment by SVGreg** [comment]
Function declaration represents interface, whereas it's content is implementation.

```swift
protocol MyAsyncProtocol {
    func doSomething() async
}

class ImpOne: MyAsyncProtocol {
    func doSomething() async {
        await useAsyncFunction() // Is Ok - using await
    }
}

class ImpTwo: MyAsyncProtocol {
    func doSomething() async {
        useSyncFunction() // Is also Ok - not to use await if not needed
    }
}

```

**Comment by pm-dev** [comment]
☝️ valid

**Comment by msrutek-paylocity** [comment]
As this has been implemented by @jkolarik-paylocity in https://github.com/realm/SwiftLint/pull/5869, this can now be closed. The rule is now opt-in. @pm-dev
